src/Groups/tom.jl

```diff
@@ -0,0 +1,613 @@
+  GAPTable::GapObj  # the GAP table of marks object
+  group::Union{GAPGroup, FinGenAbGroup}    # the underlying group, if any
+  isomorphism::Map  # isomorphism from `group` to a group in GAP
+
+  function GAPGroupTableOfMarks(G::Union{GAPGroup, FinGenAbGroup}, tom::GapObj, iso::Map)
+    return new(tom, G, iso)
+  end
+
+  function GAPGroupTableOfMarks(tom::GapObj)
+    # group and isomorphism are left undefined
+    return new(tom)
+  end
+end
+
+# access to field values via functions
+GapObj(tom::GAPGroupTableOfMarks) = tom.GAPTable
+
+function group(tom::GAPGroupTableOfMarks)
+  @req isdefined(tom, :group) "table of marks stores no group"
+  return tom.group
+end
+
+# If the table of marks stores an Oscar group `G` then
+# an isomorphism from `G` to a GAP group is stored in the field `isomorphism`.
+# Different kinds of isomorphisms can be stored,
+# depending on the type of the stored group,
+# like for character tables.
+#
+function isomorphism_to_GAP_group(tom::GAPGroupTableOfMarks)
+  @req isdefined(tom, :group) "table of marks stores no group"
+  @assert isdefined(tom, :isomorphism) "table of marks with group must store also an isomorphism"
+  return tom.isomorphism
+end
+
+
+"""
+    table_of_marks(G::GAPGroup)
+
+Return the table of marks of the finite group `G`.
+
+# Examples
+```jldoctest
+julia> show(stdout, MIME("text/plain"), table_of_marks(symmetric_group(3)))
+Table of marks of Sym(3)
+
+1: 6      
+2: 3 1    
+3: 2 . 2  
+4: 1 1 1 1
+```
+"""
+function table_of_marks(G::Union{GAPGroup, FinGenAbGroup})
+  return get_attribute!(G, :table_of_marks) do
+    iso = isomorphism_to_GAP_group(G)
+    gaptom = GAPWrap.TableOfMarks(codomain(iso))
+    return GAPGroupTableOfMarks(G, gaptom, iso)
+  end
+end
+
+# A table of marks constructed from a group is stored in this group.
+# Tables of marks from the library are cached in the dictionary
+# `tables_of_marks_by_id`,
+# in order to achieve that fetching the same table twice yields the same
+# object.
+const tables_of_marks_by_id = Dict{String, Union{GAPGroupTableOfMarks, Nothing}}()
+
+"""
+    table_of_marks(id::String)
+
+Return the table of marks
+for which `id` is an admissible name in GAP's library of tables of marks.
+If no such table is available then `nothing` is returned.
+
+# Examples
+```jldoctest
+julia> println(table_of_marks("A5"))
+table of marks of A5
+
+julia> println(table_of_marks("J5"))
+nothing
+```
+"""
+function table_of_marks(id::String)
+  # normalize `id`
+  info = GAPWrap.LibInfoCharacterTable(GapObj(id))
+  if info !== GAP.Globals.fail
+    id = string(info.firstName)
+  end
+  return get!(tables_of_marks_by_id, id) do
+    tom = GAPWrap.TableOfMarks(GapObj(id))
+    tom === GAP.Globals.fail && return nothing
+    if GAP.Globals.HasUnderlyingGroup(tom)
+      G = _oscar_group(GAPWrap.UnderlyingGroup(tom))
+      iso = isomorphism_to_GAP_group(G)
+      return GAPGroupTableOfMarks(G, tom, iso)
+    else
+      return GAPGroupTableOfMarks(tom)
+    end
```

Comment:
```suggestion
    end::GAPGroupTableOfMarks
```
to make it type stable / allow Julia to determine the return type correctly

Reply:
o.k.